airbyte-integrations/connectors/destination-jdbc/src/main/java/io/airbyte/integrations/destination/jdbc/AbstractJdbcDestination.java

```diff
@@ -0,0 +1,205 @@
+import org.slf4j.LoggerFactory;
+
+public abstract class AbstractJdbcDestination extends AbstractDestination implements Destination {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(AbstractJdbcDestination.class);
+  protected static final String COLUMN_NAME = "data";
+
+  private final String driverClass;
+  private final SQLDialect dialect;
+  private final SQLNamingResolvable namingResolver;
+  private Database databaseConnection = null;
+
+  public AbstractJdbcDestination(final String driverClass, final SQLDialect dialect, final SQLNamingResolvable namingResolver) {
+    super();
+    this.driverClass = driverClass;
+    this.dialect = dialect;
+    this.namingResolver = namingResolver;
+  }
+
+  public abstract JsonNode toJdbcConfig(JsonNode config);
+
+  @Override
+  public ConnectorSpecification spec() throws IOException {
+    // return a JsonSchema representation of the spec for the integration.
+    final String resourceString = MoreResources.readResource("spec.json");
+    return Jsons.deserialize(resourceString, ConnectorSpecification.class);
+  }
+
+  @Override
+  public AirbyteConnectionStatus check(JsonNode config) {
+    try (final Database database = getDatabase(config)) {
+      // attempt to get current schema. this is a cheap query to sanity check that we can connect to the
+      // database. `currentSchema()` is a jooq method that will run the appropriate query based on which
+      // database it is connected to.
+      database.query(this::getCurrentDatabaseName);
+
+      return new AirbyteConnectionStatus().withStatus(Status.SUCCEEDED);
+    } catch (Exception e) {
+      LOGGER.debug("Exception while checking connection: ", e);
+      return new AirbyteConnectionStatus()
+          .withStatus(Status.FAILED)
+          .withMessage("Can't connect with provided configuration.");
+    }
+  }
+
+  protected Database getDatabase(JsonNode config) {
+    final JsonNode jdbcConfig = toJdbcConfig(config);
+
+    return Databases.createDatabase(
+        jdbcConfig.get("username").asText(),
+        jdbcConfig.has("password") ? jdbcConfig.get("password").asText() : null,
+        jdbcConfig.get("jdbc_url").asText(),
+        driverClass,
+        dialect);
+  }
+
+  protected String getCurrentDatabaseName(DSLContext ctx) {
+    return ctx.select(currentSchema()).fetch().get(0).getValue(0, String.class);
+  }
+
+  @Override
+  public SQLNamingResolvable getNamingResolver() {
+    return namingResolver;
+  }
+
+  /**
+   * Return the default schema where to data in the destination if the catalog doesn't specify any
+   * location
+   *
+   * @param config The destination configuration
+   * @return default schema name where to save data
+   */
+  protected String getDefaultSchemaName(JsonNode config) {
+    if (config.has("schema")) {
+      return config.get("schema").asText();
+    } else {
+      return "public";
+    }
+  }
+
+  @Override
+  protected void connectDatabase(JsonNode config) {
```

Comment:
why do wee need both `connectDatabase` and `getDatabaseConnection`? for jdbc can't we just have `getDatabaseConnection`?

Reply:
with `connectDatabase`, you are providing the proper config data to connect to the database but when we call `getDatabaseConnection` we don't always have the config around... which is why we also store the database in a state here: https://github.com/airbytehq/airbyte/pull/1116/files#r532919911